### ISSUE TITLE
clang-tidy 22 regression: modernize-type-traits incorrectly diagnoses captured structured bindings

### ISSUE THREAD

**Issue by vogelsgesang** [llvm/llvm-project#179386]
Running `clang-tidy -checks=-*,modernize-type-traits input.cpp -- clang -std=c++23` on

```cpp
#include <utility>
using namespace std;

pair<int, int> getPair() { return make_pair(1,2); }

int test() {
   auto [a, b] = getPair();
   auto myLambda = [&]() {
      return a*2;
   };
   return myLambda();
}
```

diagnoses

```
input.cpp:9:14: warning: use c++14 style type templates [modernize-type-traits]
    9 |       return a*2 + b;
      |              ^
      |              _t
```

which is obviously incorrect.

This (or rather the non-minified repro) used to work in clang-tidy 21. Seems to be a regression in clang-tidy-22

**Comment by localspook** [comment]
Reduced (https://godbolt.org/z/WrG7hWq3K):

```cpp
namespace std {

template <typename T>
struct remove_cvref {
    using type = T;
};

} // namespace std

template <typename T>
struct S {
    std::remove_cvref<T>::type a; // NOLINT(modernize-type-traits)
};

void f() {
   auto [a] = S<int>{};
   [&] { a; }; // False positive
}
```

**Comment by localspook** [comment]
Looking at [the history of this check](https://github.com/llvm/llvm-project/commits/c57f8acee44b842f120358b49289a80ab09ff78f/clang-tools-extra/clang-tidy/modernize/TypeTraitsCheck.cpp), it seems likely that this is fallout from commit [91cdd35](https://github.com/llvm/llvm-project/commit/91cdd35008e9ab32dffb7e401cdd7313b3461892). That commit shipped in 22, so it's consistent with the regression, and I don't see other obvious candidates.

**Comment by vogelsgesang** [comment]
> Okay, I can confirm that it reproduces locally with clang-tidy 20.1.4, so it seems this isn't a 22 regression.

Your minified repro does indeed also repro already on clang-tidy 21 for me.
However, my original non-minified repro used to work with clang-tidy 21 and started failing only with clang-tidy 22
